ui/pages/MarketplaceApp.tsx

```diff
@@ -140,14 +140,48 @@ const MarketplaceApp = () => {
     }
 
     try {
+      // get the custom url from the query
       const customUrl = getQueryParamString(router.query.url);
-      const customOrigin = new URL(customUrl).origin;
-      const appOrigin = new URL(data.url).origin;
 
-      if (customOrigin === appOrigin) {
-        return customUrl;
-      } else {
-        removeQueryParam(router, 'url');
+      if (customUrl) {
+        const customOrigin = new URL(customUrl).origin;
+        const appOrigin = new URL(data.url).origin;
+        if (customOrigin === appOrigin) {
+          return customUrl;
+        } else {
+          removeQueryParam(router, 'url');
+        }
+      }
+    } catch (err) {}
+
+    try {
+      // get hash and params (using asPath to avoid conflicts with dynamic route params)
+      const [ , queryAndHash ] = router.asPath.split('?');
+      const [ queryString, hash ] = queryAndHash ? queryAndHash.split('#') : [ '', '' ];
+      const customHash = hash ? `#${ hash }` : '';
+      const customParams = new URLSearchParams(queryString);
+
+      // remove reserved params
+      [ 'url', 'action' ].forEach((param) => customParams.delete(param));
+
+      if (customParams.toString() || customHash) {
+        const targetUrl = new URL(data.url);
+        const targetParams = new URLSearchParams(targetUrl.search);
+
+        let customPath = customParams.get('path');
+        if (customPath) {
+          customPath = customPath.startsWith('/') ? customPath : `/${ customPath }`;
+          targetUrl.pathname = customPath;
+          customParams.delete('path');
+        }
+
+        customParams.forEach((value, key) => {
+          targetParams.append(key, value);
+        });
```

Comment:
Maybe we can rewrite it in a more functional way and make it more readable

```suggestion
        const targetUrl = new URL(customParams.get('path') || '', data.url);
        const targetParams = new URLSearchParams({
           ...targetUrl.search,
           ..._omit(customParams, 'path')
        });
```